### PR TITLE
[precompile] Add torch.precompile

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #187670
* #187669
* #187668
* #187667

Adds the public API: torch.precompile(fn, *args, backend=...) -> (python_code,
cache) and torch.precompile.load(python_code, cache). fn is the whole
computation (e.g. lambda m, x: m(x), or a training step that calls .backward());
nn.Module args have their params/buffers lifted to graph inputs via functional
reparametrization, so the artifact embeds NO weights -- you pass the model again
at runtime. Capture is a non-strict make_fx trace with an explicit contract,
documented in Note [precompile programming model]; the cheaply-checkable
invariants are enforced (baked constant tensors and effectful ops raise) and the
rest are the caller's responsibility.

python_code is the single source of truth for the calling convention and runs
standalone (JITing kernels); the cache is purely an acceleration consumed only by
load() (it skips JIT via an FxGraphCache hit). backend="inductor" (default)
lowers through aot_autograd.compile_to_python; backend="eager" runs the captured
ATen graph as-is for inspection. If fn ran a backward, parameter gradients are
harvested as extra graph outputs and the driver scatters them onto the runtime
model's .grad fields, accumulating like eager .backward() (only params that
actually received a gradient, so frozen params keep .grad = None).

load() reads the cache envelope with weights_only=True; the executable artifact
is the inner bytes, reconstructed by CompiledArtifact.load, which runs embedded
code -- so only load trusted caches.

Test Plan:

```
python test/test_precompile.py
```

Authored with an AI assistant.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Abandoning the stacked approach; continuing with the single PR #187258 instead.
